diff --git a/ql/pricingengines/diffusioncalculator.cpp b/ql/pricingengines/diffusioncalculator.cpp
index df88638..b8cdde3 100644
--- a/ql/pricingengines/diffusioncalculator.cpp
+++ b/ql/pricingengines/diffusioncalculator.cpp
@@ -78,17 +78,17 @@ namespace QuantLib {
         if (inputVolType == VolatilityType::ShiftedLognormal &&
             targetVolType == VolatilityType::Normal) {
             double slnVol = inputVol;
-            double nVol =
+            Real nVol =
                 convertShiftedLognormalToNormalVol(forward, strike, t, slnVol, inputDisplacement);
-            return std::make_tuple(nVol, VolatilityType::Normal, targetDisplacement);
+            return std::make_tuple(xad::value(nVol), VolatilityType::Normal, targetDisplacement);
         }
 
         if (inputVolType == VolatilityType::Normal &&
             targetVolType == VolatilityType::ShiftedLognormal) {
             double nVol = inputVol;
-            double slnVol =
+            Real slnVol =
                 convertNormalToShiftedLogNormalVol(forward, strike, t, nVol, targetDisplacement);
-            return std::make_tuple(slnVol, VolatilityType::ShiftedLognormal, targetDisplacement);
+            return std::make_tuple(xad::value(slnVol), VolatilityType::ShiftedLognormal, targetDisplacement);
         }
 
         if (inputVolType == VolatilityType::ShiftedLognormal &&
@@ -96,9 +96,9 @@ namespace QuantLib {
             !close_enough(inputDisplacement, targetDisplacement)) {
             // need to convert the vol to the new displacement
             double slnVol = inputVol;
-            double newSlnVol = convertShiftedLognormalToShiftedLognormalVol(
+            Real newSlnVol = convertShiftedLognormalToShiftedLognormalVol(
                 forward, strike, t, slnVol, inputDisplacement, targetDisplacement);
-            return std::make_tuple(newSlnVol, VolatilityType::ShiftedLognormal,
+            return std::make_tuple(xad::value(newSlnVol), VolatilityType::ShiftedLognormal,
                                    targetDisplacement);
         }
 
@@ -115,7 +115,7 @@ namespace QuantLib {
                                        double strike,
                                        double t) {
         return convertInputVolatility(outputModelType, displacement, volTS->volType(),
-                                      volTS->shift(), volTS->blackVol(t, strike), forward, strike,
+                                      xad::value(volTS->shift()), xad::value(volTS->blackVol(t, strike)), forward, strike,
                                       t);
     }
 
