diff --git a/ql/pricingengines/diffusioncalculator.cpp b/ql/pricingengines/diffusioncalculator.cpp
index df88638..560c8df 100644
--- a/ql/pricingengines/diffusioncalculator.cpp
+++ b/ql/pricingengines/diffusioncalculator.cpp
@@ -25,21 +25,21 @@
 #include <ql/pricingengines/diffusioncalculator.hpp>
 namespace QuantLib {
 
-    double convertNormalToShiftedLogNormalVol(
+    Real convertNormalToShiftedLogNormalVol(
         double forward, double strike, double ttm, double nVol, double displacement) {
         QL_REQUIRE(!close_enough(forward, 0.0), "forward is 0.0, can not convert normal vol to shifted lognormal vol");
         auto optionType = forward > strike ? Option::Type::Put : Option::Type::Call;
-        double premium = bachelierBlackFormula(optionType, strike, forward, nVol * std::sqrt(ttm));
-        double slnVol =
+        Real premium = bachelierBlackFormula(optionType, strike, forward, nVol * std::sqrt(ttm));
+        Real slnVol =
             blackFormulaImpliedStdDev(optionType, strike, forward, premium, 1.0, displacement, nVol / forward );
         return slnVol / sqrt(ttm);
     }
 
-    double convertShiftedLognormalToNormalVol(
+    Real convertShiftedLognormalToNormalVol(
         double forward, double strike, double ttm, double slnVol, double displacement) {
         auto optionType = forward > strike ? Option::Type::Put : Option::Type::Call;
-        double price = blackFormula(optionType, strike, forward, slnVol * std::sqrt(ttm), 1.0, displacement);
-        double nVol = bachelierBlackFormulaImpliedVol(optionType, strike, forward, ttm, price);
+        Real price = blackFormula(optionType, strike, forward, slnVol * std::sqrt(ttm), 1.0, displacement);
+        Real nVol = bachelierBlackFormulaImpliedVol(optionType, strike, forward, ttm, price);
         return nVol;
     }
 
@@ -49,8 +49,8 @@ namespace QuantLib {
         if (close_enough(oldDisplacement, newDisplacement))
             return slnVol;
         auto optionType = forward > strike ? Option::Type::Put : Option::Type::Call;
-        double price = blackFormula(optionType, strike, forward, slnVol, 1.0, oldDisplacement);
-        double newSlnVol =
+        Real price = blackFormula(optionType, strike, forward, slnVol, 1.0, oldDisplacement);
+        Real newSlnVol =
             blackFormulaImpliedStdDev(optionType, strike, forward, price, 1.0, newDisplacement);
         return newSlnVol / sqrt(ttm);
     }
@@ -78,7 +78,7 @@ namespace QuantLib {
         if (inputVolType == VolatilityType::ShiftedLognormal &&
             targetVolType == VolatilityType::Normal) {
             double slnVol = inputVol;
-            double nVol =
+            Real nVol =
                 convertShiftedLognormalToNormalVol(forward, strike, t, slnVol, inputDisplacement);
             return std::make_tuple(nVol, VolatilityType::Normal, targetDisplacement);
         }
@@ -86,7 +86,7 @@ namespace QuantLib {
         if (inputVolType == VolatilityType::Normal &&
             targetVolType == VolatilityType::ShiftedLognormal) {
             double nVol = inputVol;
-            double slnVol =
+            Real slnVol =
                 convertNormalToShiftedLogNormalVol(forward, strike, t, nVol, targetDisplacement);
             return std::make_tuple(slnVol, VolatilityType::ShiftedLognormal, targetDisplacement);
         }
@@ -96,7 +96,7 @@ namespace QuantLib {
             !close_enough(inputDisplacement, targetDisplacement)) {
             // need to convert the vol to the new displacement
             double slnVol = inputVol;
-            double newSlnVol = convertShiftedLognormalToShiftedLognormalVol(
+            Real newSlnVol = convertShiftedLognormalToShiftedLognormalVol(
                 forward, strike, t, slnVol, inputDisplacement, targetDisplacement);
             return std::make_tuple(newSlnVol, VolatilityType::ShiftedLognormal,
                                    targetDisplacement);
@@ -114,8 +114,10 @@ namespace QuantLib {
                                        double forward,
                                        double strike,
                                        double t) {
+        Real shift = volTS->shift();
+        Real blackVol = volTS->blackVol(t, strike);
         return convertInputVolatility(outputModelType, displacement, volTS->volType(),
-                                      volTS->shift(), volTS->blackVol(t, strike), forward, strike,
+                                      shift, blackVol, forward, strike,
                                       t);
     }
 
diff --git a/ql/pricingengines/diffusioncalculator.hpp b/ql/pricingengines/diffusioncalculator.hpp
index fa2844a..736239e 100644
--- a/ql/pricingengines/diffusioncalculator.hpp
+++ b/ql/pricingengines/diffusioncalculator.hpp
@@ -69,10 +69,10 @@ namespace QuantLib {
                            double t);
 
 
-    double convertNormalToShiftedLogNormalVol(
+    Real convertNormalToShiftedLogNormalVol(
         double forward, double strike, double ttm, double nVol, double displacement);
 
-    double convertShiftedLognormalToNormalVol(
+    Real convertShiftedLognormalToNormalVol(
         double forward, double strike, double ttm, double slnVol, double displacement);
 
 
