diff --git a/ql/pricingengines/vanilla/analyticeuropeanengine.cpp b/ql/pricingengines/vanilla/analyticeuropeanengine.cpp
index 62dee1b..73c3187 100644
--- a/ql/pricingengines/vanilla/analyticeuropeanengine.cpp
+++ b/ql/pricingengines/vanilla/analyticeuropeanengine.cpp
@@ -89,8 +89,8 @@ namespace QuantLib {
         Real forwardPrice = s0 * dividendDiscount / riskFreeDiscountForFwdEstimation;
 
         auto [variance, volType, displacement] = convertInputVariance(
-            modelType_, displacement_, *process_->blackVolatility(), forwardPrice, payoff->strike(),
-            process_->blackVolatility()->timeFromReference(arguments_.exercise->lastDate()));
+            modelType_, xad::value(displacement_), *process_->blackVolatility(), xad::value(forwardPrice), xad::value(payoff->strike()),
+            xad::value(process_->blackVolatility()->timeFromReference(arguments_.exercise->lastDate())));
         
         DiffusionCalculator black(payoff, forwardPrice, std::sqrt(variance), df, volType, displacement);
 
