diff --git a/qle/models/parametrization.hpp b/qle/models/parametrization.hpp
index 5effd88..311395b 100644
--- a/qle/models/parametrization.hpp
+++ b/qle/models/parametrization.hpp
@@ -111,7 +111,7 @@ inline Time Parametrization::tr(const Time t) const { return t > 0.5 * h_ ? t +
 
 inline Time Parametrization::tl(const Time t) const { return std::max(t - 0.5 * h_, 0.0); }
 
-inline Time Parametrization::tr2(const Time t) const { return t > h2_ ? t + h2_ : 2.0 * h2_; }
+inline Time Parametrization::tr2(const Time t) const { return t > h2_ ? Real(t + h2_) : Real(2.0 * h2_); }
 
 inline Time Parametrization::tm2(const Time t) const { return t > h2_ ? t : h2_; }
 
