diff --git a/qle/models/defaultableequityjumpdiffusionmodel.hpp b/qle/models/defaultableequityjumpdiffusionmodel.hpp
index 47ec2dc..d89cd1d 100644
--- a/qle/models/defaultableequityjumpdiffusionmodel.hpp
+++ b/qle/models/defaultableequityjumpdiffusionmodel.hpp
@@ -186,7 +186,7 @@ private:
     mutable Real totalBlackVariance_ = 1.0;
 
     // step size to compute r, q with finite differences
-    static constexpr Real fh_ = 1E-4;
+    static inline const Real fh_ = 1E-4;
 };
 
 } // namespace QuantExt
diff --git a/qle/pricingengines/mcmultilegbaseengine.hpp b/qle/pricingengines/mcmultilegbaseengine.hpp
index f6aa734..5fac2c3 100644
--- a/qle/pricingengines/mcmultilegbaseengine.hpp
+++ b/qle/pricingengines/mcmultilegbaseengine.hpp
@@ -151,7 +151,7 @@ public:
     // results, these are read from derived engines
     mutable Real resultUnderlyingNpv_, resultValue_;
 
-    static constexpr Real tinyTime = 1E-10;
+    static inline const Real tinyTime = 1E-10;
 
     // data structure storing info needed to generate the amount for a cashflow
     struct CashflowInfo {
