diff --git a/qle/pricingengines/blackmultilegoptionengine.cpp b/qle/pricingengines/blackmultilegoptionengine.cpp
index 18b21cb..a780e32 100644
--- a/qle/pricingengines/blackmultilegoptionengine.cpp
+++ b/qle/pricingengines/blackmultilegoptionengine.cpp
@@ -195,7 +195,7 @@ void BlackMultiLegOptionEngineBase::calculate() const {
                     cfInfo.back().discountFactor = discountCurve_->discount(c->date());
                     // mid cpn exercise assumes short coupon is paid
                     cfInfo.back().amount =
-                        c->amount() * payer * (midCouponExercise_ ? cfInfo.back().accrual / c->accrualPeriod() : 1.0);
+                        c->amount() * payer * (midCouponExercise_ ? Real(cfInfo.back().accrual / c->accrualPeriod()) : Real(1.0));
                     cfInfo.back().payDate = c->date();
                 }
 
@@ -255,7 +255,7 @@ void BlackMultiLegOptionEngineBase::calculate() const {
         }
     }
     Real weightedFixedRate =
-        QuantLib::close_enough(weightedFixedRateDenom, 0.0) ? 0.0 : weightedFixedRateNom / weightedFixedRateDenom;
+        QuantLib::close_enough(weightedFixedRateDenom, 0.0) ? Real(0.0) : Real(weightedFixedRateNom / weightedFixedRateDenom);
 
     // determine the pv-weighted floating margin, the floating bps, the floating npv
 
@@ -271,8 +271,8 @@ void BlackMultiLegOptionEngineBase::calculate() const {
         }
     }
     Real weightedFloatingSpread = QuantLib::close_enough(weightedFloatingSpreadDenom, 0.0)
-                                      ? 0.0
-                                      : weightedFloatingSpreadNom / weightedFloatingSpreadDenom;
+                                      ? Real(0.0)
+                                      : Real(weightedFloatingSpreadNom / weightedFloatingSpreadDenom);
 
     // determine the simple cf npv
 
@@ -285,12 +285,12 @@ void BlackMultiLegOptionEngineBase::calculate() const {
 
     // determine the fair swap rate
 
-    Real atmForward = close_enough(fixedBps, 0.0) ? 0.0 : -floatingNpv / fixedBps;
+    Real atmForward = close_enough(fixedBps, 0.0) ? Real(0.0) : Real(-floatingNpv / fixedBps);
 
     // determine the spread correction
 
     Real spreadCorrection =
-        weightedFloatingSpread * (QuantLib::close_enough(fixedBps, 0.0) ? 0.0 : std::abs(floatingBps / fixedBps));
+        weightedFloatingSpread * (QuantLib::close_enough(fixedBps, 0.0) ? Real(0.0) : Real(std::abs(floatingBps / fixedBps)));
 
     Real effectiveAtmForward = atmForward - spreadCorrection;
     Real effectiveFixedRate = weightedFixedRate - spreadCorrection;
@@ -300,7 +300,7 @@ void BlackMultiLegOptionEngineBase::calculate() const {
     Real simpleCfCorrection = 0.0;
     for (auto const& c : cfInfo) {
         if (c.fixedRate == Null<Real>() && c.floatingSpread == Null<Real>()) {
-            simpleCfCorrection += QuantLib::close_enough(fixedBps, 0.0) ? 0.0 : c.amount * c.discountFactor / fixedBps;
+            simpleCfCorrection += QuantLib::close_enough(fixedBps, 0.0) ? Real(0.0) : Real(c.amount * c.discountFactor / fixedBps);
         }
     }
 
diff --git a/qle/pricingengines/cbomcengine.cpp b/qle/pricingengines/cbomcengine.cpp
index c69e6b2..095a277 100644
--- a/qle/pricingengines/cbomcengine.cpp
+++ b/qle/pricingengines/cbomcengine.cpp
@@ -46,8 +46,8 @@ namespace QuantExt {
         }
 
         Real ccyDis = (iFlows[ccy][j].flow_ > 0 ?
-                       iFlows[ccy][j].discountedFlow_ / iFlows[ccy][j].flow_:
-                       0.0);
+                       Real(iFlows[ccy][j].discountedFlow_ / iFlows[ccy][j].flow_):
+                       Real(0.0));
 
         // Accrued Interest
 
@@ -80,8 +80,8 @@ namespace QuantExt {
         Currency ccy = arguments_.ccy;
 
         Real ccyDis = (iFlows[ccy][j].flow_ > 0 ?
-                       iFlows[ccy][j].discountedFlow_ / iFlows[ccy][j].flow_:
-                       0.0);
+                       Real(iFlows[ccy][j].discountedFlow_ / iFlows[ccy][j].flow_):
+                       Real(0.0));
 
         //IC and OC
 
@@ -120,8 +120,8 @@ namespace QuantExt {
         Currency ccy = arguments_.ccy;
 
         Real ccyDis = (pFlows[ccy][j].flow_ > 0 ?
-                       pFlows[ccy][j].discountedFlow_ / pFlows[ccy][j].flow_:
-                       0.0);
+                       Real(pFlows[ccy][j].discountedFlow_ / pFlows[ccy][j].flow_):
+                       Real(0.0));
 
         //Principal Waterfall
 
@@ -352,14 +352,14 @@ namespace QuantExt {
 
                 for(size_t d = 0; d < dates.size(); d++){
 
-                    double cf1 = 0.0;
-                    double cf2 = 0.0;
-                    double if1 = 0.0;
-                    double if2 = 0.0;
-                    double pf1 = 0.0;
-                    double pf2 = 0.0;
+                    Real cf1 = 0.0;
+                    Real cf2 = 0.0;
+                    Real if1 = 0.0;
+                    Real if2 = 0.0;
+                    Real pf1 = 0.0;
+                    Real pf2 = 0.0;
                     //Cash pf(0.0, 0.0);
-                    double bn = 0.0;
+                    Real bn = 0.0;
 
                     for(auto& basketCcy : basketCurrency){
                         cf1 += arguments_.basket->convert(cf_full[basketCcy][d].flow_, basketCcy, dates[d]);
@@ -396,8 +396,8 @@ namespace QuantExt {
                 //Back out discountfactors 
 
                 Real intCcyDis = (iFlows[ccy][j].flow_ > 0 ?
-                               iFlows[ccy][j].discountedFlow_ / iFlows[ccy][j].flow_ :
-                               0.0);
+                               Real(iFlows[ccy][j].discountedFlow_ / iFlows[ccy][j].flow_) :
+                               Real(0.0));
 
                 /**************************************************************
                  * check flows add up
