diff --git a/QuantExt/qle/math/randomvariable.hpp b/QuantExt/qle/math/randomvariable.hpp
index 7582b14..8eb0133 100644
--- a/QuantExt/qle/math/randomvariable.hpp
+++ b/QuantExt/qle/math/randomvariable.hpp
@@ -362,7 +362,7 @@ inline void RandomVariable::set(const Size i, const Real v) {
             return;
         }
     }
-    data_[i] = v;
+    data_[i] = static_cast<double>(v);
 }
 
 inline Real RandomVariable::operator[](const Size i) const {
