diff --git a/qle/pricingengines/analyticlgmcdsoptionengine.cpp b/qle/pricingengines/analyticlgmcdsoptionengine.cpp
index bcea0ee..0f23924 100644
--- a/qle/pricingengines/analyticlgmcdsoptionengine.cpp
+++ b/qle/pricingengines/analyticlgmcdsoptionengine.cpp
@@ -73,11 +73,11 @@ void AnalyticLgmCdsOptionEngine::calculate() const {
         t_[i + 1] = yts->timeFromReference(cpn->date());
         Real mid = (t_[i] + t_[i + 1]) / 2.0;
         if (arguments_.swap->settlesAccrual()) {
-            Real accStartTime = i == 0 ? yts->timeFromReference(cpn->accrualStartDate()) : t_[i];
+            Real accStartTime = i == 0 ? Real(yts->timeFromReference(cpn->accrualStartDate())) : Real(t_[i]);
             // mid > accStartTime practically always the case?
-            accrualSettlementAmount = mid > accStartTime ? swapSpread * cpn->accrualPeriod() * (mid - accStartTime) /
-                                                               (t_[i + 1] - accStartTime)
-                                                         : 0.0;
+            accrualSettlementAmount = mid > accStartTime ? Real(swapSpread * cpn->accrualPeriod() * (mid - accStartTime) /
+                                                               (t_[i + 1] - accStartTime))
+                                                         : Real(0.0);
         }
         C[i] = ((1.0 - recoveryRate_) - accrualSettlementAmount) * yts->discount(mid) / yts->discount(tex_);
         D[i] = swapSpread * cpn->accrualPeriod() * yts->discount(t_[i + 1]) / yts->discount(tex_);
