diff --git a/qle/pricingengines/blackswaptionenginedeltagamma.hpp b/qle/pricingengines/blackswaptionenginedeltagamma.hpp
index a8a046b..c858c1c 100644
--- a/qle/pricingengines/blackswaptionenginedeltagamma.hpp
+++ b/qle/pricingengines/blackswaptionenginedeltagamma.hpp
@@ -457,11 +457,11 @@ template <class Spec> void BlackStyleSwaptionEngineDeltaGamma<Spec>::calculate()
             Matrix result(2 * n, 2 * n, 0.0);
             for (Size i = 0; i < 2 * n; ++i) {
                 bool iIsDsc = i < n;
-                Real Fi = iIsDsc ? (mu_sd[i] / annuity - mu * A_s[i] / (annuity * annuity)) : (mu_sf[i - n] / annuity);
+                Real Fi = iIsDsc ? Real(mu_sd[i] / annuity - mu * A_s[i] / (annuity * annuity)) : Real(mu_sf[i - n] / annuity);
                 for (Size j = 0; j <= i; ++j) {
                     bool jIsDsc = j < n;
                     Real Fj =
-                        jIsDsc ? (mu_sd[j] / annuity - mu * A_s[j] / (annuity * annuity)) : (mu_sf[j - n] / annuity);
+                        jIsDsc ? Real(mu_sd[j] / annuity - mu * A_s[j] / (annuity * annuity)) : Real(mu_sf[j - n] / annuity);
                     Real Fij = mu_ss[i][j] / annuity;
                     if (iIsDsc) {
                         if (jIsDsc) {
