diff --git a/qle/pricingengines/indexcdstrancheengine.cpp b/qle/pricingengines/indexcdstrancheengine.cpp
index 2eaaf8b..8f1c424 100644
--- a/qle/pricingengines/indexcdstrancheengine.cpp
+++ b/qle/pricingengines/indexcdstrancheengine.cpp
@@ -85,14 +85,14 @@ void IndexCdsTrancheEngine::calculate() const {
     results_.premiumValue = 0.0;
     results_.protectionValue = 0.0;
     Real inceptionTrancheNotional = arguments_.basket->trancheNotional();
-    std::vector<double> effectiveNotionals;
+    std::vector<Real> effectiveNotionals;
     std::vector<Date> defaultDates;
 
-    std::vector<double> defaultDiscountFactors;
-    std::vector<double> premiumAccrualPeriods;
-    std::vector<double> premiumDiscountFactors;
-    std::vector<double> zeroRecoveryExpectedLoss;
-    std::vector<double> accrualsDefault;
+    std::vector<Real> defaultDiscountFactors;
+    std::vector<Real> premiumAccrualPeriods;
+    std::vector<Real> premiumDiscountFactors;
+    std::vector<Real> zeroRecoveryExpectedLoss;
+    std::vector<Real> accrualsDefault;
     // Value the premium and protection leg.
     effectiveNotionals.push_back(inceptionTrancheNotional); // ensure at least one effective notional theree
     for (Size i = 0; i < arguments_.normalizedLeg.size(); i++) {
